ULog File Format Document Improvement (#1887)
PX4/PX4-user_guide: en/dev_log/ulog_file_format.md

# ULog File Format

ULog (Universal-Log) is the file format used for logging messages. The format is self-describing, i.e. it contains the format and [uORB](../middleware/uorb.md) message types that are logged.

> This document is meant to be the ULog File Format Spec Documentation, especially for anyone who is interested in developing a ULog Parser / Application and needs to decode ULog file.

Reviewer: FWIW Just as easy to do these kinds of things as suggestions.
```suggestion
This document is meant to be the ULog File Format Spec Documentation.
It is intended especially for anyone who is interested in writing a ULog parser / serializer and needs to decode / encode files.
```

# ULog File Format

> ULog (Universal-Log) is the file format used for logging messages. The format is self-describing, i.e. it contains the format and [uORB](../middleware/uorb.md) message types that are logged.

Reviewer 2: @junwoo091400 as I told you I'm not sure if that's the exact origin
```suggestion
ULog is the file format used for logging messages. The format is self-describing, i.e. it contains the format and [uORB](../middleware/uorb.md) message types that are logged.
```

# ULog File Format

ULog is the file format used for logging system data. The format is self-describing, i.e. it contains the format and [uORB](../middleware/uorb.md) message types that are logged.

::: note
The [system logger](../modules/modules_system.md#logger) allows the default set of logged topics to be replaced with a custom set that is defined on an SD Card.

> See [Logging > SD Card Configuration](../dev_log/logging#sd-card-configuration) for more information.

Author: Removed! And yes it's irrelevant for the ULog File Format Spec!

:::

> ULog commonly logs uORB topics related to (but not limited to) the following sources:

Author: Changes applied!

# ULog File Format

ULog is the file format used for logging system data.
The format is self-describing, i.e. it contains the format and [uORB](../middleware/uorb.md) message types that are logged.

::: note
The [system logger](../modules/modules_system.md#logger) allows the default set of logged topics to be replaced with a custom set that is defined on an SD Card.
See [Logging > SD Card Configuration](../dev_log/logging.html#sd-card-configuration) for more information.
:::

ULog commonly logs uORB topics related to (but not limited to) the following sources:
- **Device inputs:** Sensors, RC input, etc.
- **Internal states:** CPU load, attitude, EKF state, etc.

> - **Messages:** MAVLink communications, `PX4_INFO()`, Custom `printf` statements, and so on.

Author: Gotcha, will do that

> - **Messages:** MAVLink communications, `PX4_INFO()`, Custom `printf` statements, and so on.

Author: @bkueng Is it correct to say that Messages part (MAVLink message, PX4_INFO, etc) are also a uORB topic? Or how should we phrase this part?:

>  ULog commonly logs uORB topics related to (but not limited to) the following sources:

> - **Messages:** MAVLink communications, `PX4_INFO()`, Custom `printf` statements, and so on.

Reviewer 2: Hamish is correct, it's either the tagged or non-tagged string messages.
```
**String messages:** `printf` statements, including `PX4_INFO()` and `PX4_ERR()`.
```

> - **Messages:** MAVLink communications, `PX4_INFO()`, Custom `printf` statements, and so on.

Author: Thank you!

> - **Messages:** MAVLink communications, `PX4_INFO()`, Custom `printf` statements, and so on.

Reviewer: @bkueng I've been playing with a new MAVLink battery message for PX4 dev summit. The updated uorb topic shows up in the log, but I'm not seeing any mavlink messages. Do these have to be explicitly enabled? If so, how?

> - **Messages:** MAVLink communications, `PX4_INFO()`, Custom `printf` statements, and so on.

Reviewer 2: What do you mean by mavlink messages do not show up? Where do you expect them?

> - **Messages:** MAVLink communications, `PX4_INFO()`, Custom `printf` statements, and so on.

Reviewer: I was confused. I was assuming there might be magic that automatically logs all mavlink messages. But that is not what this says - it logs uorb topics.

ULog is the file format used for logging system data.

ULog is **self-describing** : It contains the format and message types that are logged.

::: note

> The [logger](../dev_log/logging.md#sd-card-configuration) allows the specifying a custom *default set* of logged topics with the `logger_topics.txt` file

Author: Thanks! I checked the change

|double             |  8            |
|bool, char         |  1            |

> Additionally the types can be used as an array : eg. `float[5]`. In case of logging of strings : (`char[length]`), it usually does not contain the termination NULL character `'\0'` at the end.

Reviewer: Note again, why have all that space around the colon? `an array : eg.`.
The first sentence is an improvement, the second is not. I will fix.

String comparisons are case sensitive.

> ## ULog structure

Reviewer: Ignore that, it makes no difference here.

All I was suggesting (now fixed) is `## ULog Structure`

> ULog consists of the following three sections:

Reviewer: You're talking about a file here, so you wouldn't normally rename to "ulog", and if you do, you'd say "the ULog file format".

SO this would reasonably change to
```
The ULog file format consists of the following three sections:
```
or 
```
The file consists of the following three sections:
```

Author: I thought ULog itself is a file format, therefore there is no need to state it again (even the document's name is `ulog_file_format.md`)

Reviewer: YOu could be right - not sure it is defined as such though. But would you say "Zip consists of" or "Mission plan consists of"?

Author: Hmm yea that wouldn't make much sense 🤔

|double             |  8            |
|bool, char         |  1            |

Additionally the types can be used as an array: e.g. `float[5]`.

Strings (`char[length]`) do not usually contain the termination NULL character `'\0'` at the end.

> String comparisons are case sensitive.

Author: @bkueng I am still wondering if this part is still relevant / necessary for the User Doc. Could we bring a contextual example that shows when 'case sensitie-ness' matters?

Is it used in Flight Review? ulog python library? etc.

Reviewer 2: It's relevant, don't just remove things. This doc is meant to be a semi-formal file format spec, that can be used to implement a file reader or writer. Please keep that in mind.

----------------------
```

> Description on each section is explained below.

Reviewer: Note, this says in English "below I will explain the description" not "below I will explain the section".

Author: Ohhh yes I am Explaining the Description :rofl:, my english is :camel:

Author: Thank you for the catch!

### Definition & Data Section Message Header

The _Definitions and Data_ sections contain a number of **messages**. Each message is preceded by this header:

```c
struct message_header_s {
  uint16_t msg_size;
  uint8_t msg_type;
};
```

- `msg_size` is the size of the message in bytes without the header.
- `msg_type` defines the content, and is a single character.

:::note
Message sections below are prefixed with the character that corresponds to it's `msg_type`.
:::

### Definitions Section

The definitions section contains basic information such as software version, message format, initial parameter values, and so on.

The message types in this section are:
1. [Flag Bits](#b-flag-bits-message)
2. [Format Definition](#f-format-message)
3. [Information](#i-information-message)
4. [Multi Information](#m-multi-information-message)
5. [Parameter](#p-parameter-message)
6. [Default Parameter](#q-default-parameter-message)

#### 'B': Flag Bits Message

:::note
This message must be the **first message** right after the header section, so that it has a fixed constant offset from the start of the file!
:::

> This message provides an information to the log parser whether the log is parsable or not.

Reviewer 2: ```suggestion
This message provides information to the log parser whether the log is parsable or not.
```

```c
struct ulog_message_flag_bits_s {
  struct message_header_s header; // msg_type = 'B'
  uint8_t compat_flags[8];
  uint8_t incompat_flags[8];
  uint64_t appended_offsets[3]; // file offset(s) for appended data if appending bit is set
};
```

- `compat_flags`: compatible flag bits
  - These flags indicate the presence of features in the log file that are compatible with any ULog parser.
  - `compat_flags[0]`: *DEFAULT_PARAMETERS* (Bit 0): if set, the log contains [default parameters message](#q-default-parameter-message)

  The rest of the bits are currently not defined and must be set to 0.
  These bits can be used for future ULog changes that are compatible with existing parsers.
  For example, adding a new message type can be indicated by defining a new bit in the standard, and existing parsers will ignore the new message type.
  It means parsers can just ignore the bits if one of the unknown bits is set.
- `incompat_flags`: incompatible flag bits.
  - `incompat_flags[0]`: *DATA_APPENDED* (Bit 0): if set, the log contains appended data and at least one of the `appended_offsets` is non-zero.
  
  The rest of the bits are currently not defined and must be set to 0.

>   This can be used to introduce breaking changes that existing parsers cannot handle. For example, when an old ULog Parser that didn't have the concept of *DATA_APPENDED* reads the newer ULog, it would stop parsing the log as the log will contain out-of-spec messages / concepts.

Reviewer 2: ```suggestion
  This can be used to introduce breaking changes that existing parsers cannot handle. For example, when an old ULog parser that didn't have the concept of *DATA_APPENDED* reads the newer ULog, it would stop parsing the log as the log will contain out-of-spec messages / concepts.
```

  If a parser finds any of these bits set that isn't specified, it must refuse to parse the log.

> - `appended_offsets`: File offset for appended data.

Reviewer 2: ```suggestion
- `appended_offsets`: File offset (0-based) for appended data.
```

`uint32_t ver_os_release` | OS version (see below) |  0x010401ff
`char[value_len] sys_toolchain` | Toolchain Name |  "GNU GCC"
`char[value_len] sys_toolchain_ver` | Toolchain Version |  "6.2.1"
`char[value_len] sys_mcu` | Chip name and revision |  "STM32F42x, rev A"
`char[value_len] sys_uuid` | Unique identifier for vehicle (eg. MCU ID) |  "392a93e32fa3"...
`char[value_len] log_type` | Type of the log (full log if not specified) | "mission"
`char[value_len] replay` | File name of replayed log if in replay mode | "log001.ulg"
`int32_t time_ref_utc` | UTC Time offset in seconds | -3600 |

:::note
`value_len` represents the data size of the `value`. This is described in the `key`.
:::

- The format of `ver_sw_release` and `ver_os_release` is: 0xAABBCCTT, where AA is **major**, BB is **minor**, CC is patch and TT is the **type**.
  - **Type** is defined as following: `>= 0`: development, `>= 64`: alpha version, `>= 128`: beta version, `>= 192`: RC version, `== 255`: release version.
  - For example, `0x010402FF` translates into the release version v1.4.2.

This message can also be used in the Data section (this is however the preferred section).

#### 'M': Multi Information Message

> Multi information message serves the same purpose as the information message, but for cases where the `value` length exceeds  (parsers store them as a list).

Reviewer 2: ```suggestion
Multi information message serves the same purpose as the information message, but for long messages or multiple messages with the same key.
```

----------------------
```

A description of each section is provided below.

### Header Section

The header is a fixed-size section and has the following format (16 bytes):

```
----------------------------------------------------------------------
| 0x55 0x4c 0x6f 0x67 0x01 0x12 0x35 | 0x01         | uint64_t       |
| File magic (7B)                    | Version (1B) | Timestamp (8B) |
----------------------------------------------------------------------
```

> - **File Magic (7 Bytes):** File type indicator that reads "ULog"

Reviewer: Note the two characters at the end. THese don't render in gitbook, and Notepad shows them as: 
![image](https://user-images.githubusercontent.com/5368500/169185672-cddf9e1f-0846-4485-a24c-36669035aa1d.png)

Is there another way to describe this - e.g. "..that reads "ULogXY, where XY is [whatever it is]"

Author: > e.g. "..that reads "ULogXY, where XY is [whatever it is]"

I can say "ULogXYZ, where XYZ are the magic 3 bytes: 0x01 0x12 0x35"

Reviewer: That works.

- **Version (1 Byte):** File format version (currently 1)
- **Timestamp (8 Bytes):** `uint64_t` integer that denotes when the logging started in microseconds.

### Definition & Data Section Message Header

> The _Definitions and Data_ section consists of a stream of **messages** that always starts with this header: 

Reviewer: Are you saying that:
- every message in both sections starts with this header
- before all the stream of messages for the definitions and data this header is displayed
- before the stream of messages for the definitions section this header is display and also it is displayed again before the stream for data?

This sentence will need to change based on which one of the above is true.

Author: > * every message in both sections starts with this header

This is the correct case. I think I can change it to:

The _Definitions and Data_ section consists of **messages**, and each message always starts with this header:

Reviewer: Do you mean "starts with" or is "preceded by":
```suggestion
The _Definitions and Data_ sections contain a number of **messages**. Each message is preceded by this header: 
```

	uint16_t msg_size;
	uint8_t msg_type
};
```

- `msg_size` is the size of the message in bytes without the header.
- `msg_type` defines the content, and is a single alphabet character.

:::note
Message types described below are noted with the single alphabet `msg_type` characteristic in the beginning.
:::

### Definitions Section

The definitions section contains basic information such as software version, uORB Topic format, initial parameter values, and so on.

The message types in this section are:
1. [Flag Bits](#b--flag-bits-message)
2. [Format Definition](#f--format-definition-message)
3. [Information](#i--information-message)
4. [Multi Information](#m--multi-information-message)
5. [Parameter](#p--parameter-message)
6. [Default Parameter](#q--default-parameter-message)

#### 'B': Flag Bits Message

:::note
This message must be the **first message** right after the header section, so that it has a fixed constant offset from the start of the file!
:::

```
struct ulog_message_flag_bits_s {
  struct message_header_s;
  uint8_t compat_flags[8];
  uint8_t incompat_flags[8];
  uint64_t appended_offsets[3]; ///< file offset(s) for appended data if appending bit is set
};
```

> - `compat_flags`: compatible flag bits.

Author: @bkueng Is there a better way to phrase this? Since I am considering explaining this part like:
> Compatible flags indicate if the log includes any special message that is compatible with the existing parsers is in the uLog

And the `incompat_flags` part like:
> Incompatible flags indicate whether any special message that is incompatible with the existing parsers is in the uLog

But I am not sure if I understand this part correctly (Also out of 64 bits, only 1 bit is being used for now :thinking:), could you give your input on this?

  These bits can be used for future ULog changes that are compatible with existing parsers.
  It means parsers can just ignore the bits if one of the unknown bits is set.
- `incompat_flags`: incompatible flag bits.
  The LSB bit of index 0 is set to one if the log contains appended data and at least one of the `appended_offsets` is non-zero.
  All other bits are undefined and must be set to 0.
  If a parser finds one of these bits set, it must refuse to parse the log.
  This can be used to introduce breaking changes that existing parsers cannot handle.
- `appended_offsets`: File offsets (0-based) for appended data.
  If no data is appended, all offsets must be zero.
  This can be used to reliably append data for logs that may stop in the middle of a message.

  A process appending data should do:
  - set the relevant `incompat_flags` bit,
  - set the first `appended_offsets` that is 0 to the length of the log file,
  - then append any type of messages that are valid for the Data section.

It is possible that there are more fields appended at the end of this message in future ULog specifications.
This means a parser must not assume a fixed length of this message.

> If the message is longer than expected (currently 40 bytes), the exceeding bytes must just be ignored.

Author: This part is still very vague / hard to understand

Reviewer: Yes, that is odd wording. @bkueng Is this what is meant? 
```suggestion
It is possible that there are more fields appended at the end of this message in future ULog specifications.
This means a parser must not assume a fixed length of this message.
If the message is longer than expected (currently 40 bytes), any additional bytes must be ignored/discarded.
```

Reviewer 2: Yes

#### 'F': Format Definition Message

Format definition that can be logged or used in another format definitions as a nested type.

```
struct message_format_s {
  struct message_header_s header;
  char format[header.msg_size];
};
```

* `format` is a plain-text string with the following format: `message_name:field0;field1;`

There can be an arbitrary amount of fields (minimum 1), separated by `;`.

A field has the format: `type field_name` or `type[array_length] field_name` for arrays (only fixed size arrays are supported).

`type` is one of the basic binary types or a `message_name` of another format definition (nested usage).
A type can be used before it's defined.
There can be arbitrary nesting but no circular dependencies.

Some field names are special:
- `timestamp`: every [Subscription Message](#a--subscription-message) must include a timestamp field
  - Its type can be: `uint64_t` (currently the only one used), `uint32_t`, `uint16_t` or `uint8_t`.
  - The unit is always microseconds, except for in `uint8_t` it's milliseconds.

>   - A log writer must make sure to log messages often enough to be able to detect wrap-arounds and a log reader must handle wrap-arounds (and take into account dropouts).

Author: @bkueng I don't understand this part. Could you explain?

Reviewer: What it means is that the size of the timestamp isn't infinite, and it will rollover back to zero after a while. You can detect this is you log with some frequency, but if your log length is too infrequent you won't know if your latest item is actually at timestamp X or timestamp X + the interval for the whole timer to cycle once.

Author: Added to the doc!

In practice, we include the `msg_size` and `msg_type` in each message definitions as the first 2 members, so the header `message_header_s` never gets used directly in most cases.

:::note
Messages described below are prefixed with the single alphabet that corresponds to it's `msg_type` in the beginning.
:::

### Definitions Section

Variable length section, contains version information, format definitions, and (initial) parameter values.
The definitions section contains basic information such as software version, uORB Topic format, initial parameter values, and so on.

The message types in this section are:
1. [Flag Bits](#b-flag-bits-message)
2. [Format Definition](#f-format-message)
3. [Information](#i-information-message)
4. [Multi Information](#m-multi-information-message)
5. [Parameter](#p-parameter-message)
6. [Default Parameter](#q-default-parameter-message)

#### 'B': Flag Bits Message

:::note
This message must be the **first message** right after the header section, so that it has a fixed constant offset from the start of the file!
:::

The Definitions and Data sections consist of a stream of messages.
Each starts with this header:
```c

> struct message_header_s {

Reviewer 2: This was actually better for understanding, it makes things more clear and explicit. The fact that PX4 does it differently, isn't really relevant here.

Author: Gotcha 👌. One motive behind changing to this format was to expose the `msg_type` byte, and make it clear that each messages had their own type assigned 🤔. Will add it as a separate text!

#### 'I': Information Message

The Information message defines a dictionary type definition `key` : `value` pair for any information, including but not limited to Hardware version, Software version, Build toolchain for the software, etc.

```c
struct ulog_message_info_header_s {
  uint16_t msg_size;
  uint8_t msg_type = 'I';
  uint8_t key_len;
  char key[255];
};
```

- `key_len`: Length of the key value
- `key`: String with the `key` and `value` information combined: e.g. `char[5] sys_toolchain_ver9.4.0`
  - The `key` part has the format identical as the field in the [Format message](#f-format-message) (can also be a custom type), but consists of only a single field without ending `;`, eg. `char[5] sys_toolchain_ver`
  - The `value` contains the data corresponding to the `key`, and is written right after the `key`, e.g. `9.4.0`.

:::note
A key : value pair defined in the Information message should be unique. Meaning there shouldn't be more than one definition with the same key value!
:::

Parsers can store information messages as a dictionary.

Predefined information messages are:

key | Description | Example for value
--- | ---  | ---

> `char[3] sys_name` | Name of the system |  "PX4"

Reviewer 2: I find this more confusing than before. The key is generic, but now it's tied to the example

Author: I found the `value_len` part confusing when I first encountered the doc. Is there a better way to phrase it?

Since currently with the example, it is very clear that this is how you can use the key : value pair, but it's also not generic anymore. Maybe add more text above to state that it can be generic?

Reviewer: @bkueng I think you are saying that the values @junwoo091400 has suggested are length values specific to PX4 implementation, and that some other flight stack using this spec might use different numbers. Is that right?

There are two possible fixes. 
1. We use `value_len` but we have a note at the bottom stating (something like):
   > `value_len` represents the data size required for storing the key. This will be key and system specific.
2. Use @junwoo091400 values, but at the end have a note:
  > The values for the value data-size in the table above are those used by PX4. Other systems may use other values.
  
I tend to think option 1 is better for a "generic spec" - because with option 2 people are more likely just to see the values and assume they can be hard coded (i.e. not read the spec properly). So I'd revert and use option "1" (assuming Beat agrees)

In practice, we include the `msg_size` and `msg_type` in each message definitions as the first 2 members of different message definitions.

:::note
Message sections below are prefixed with the character that corresponds to it's `msg_type`!
:::

### Definitions Section

The definitions section contains basic information such as software version, uORB Topic format, initial parameter values, and so on.

The message types in this section are:
1. [Flag Bits](#b-flag-bits-message)
2. [Format Definition](#f-format-message)
3. [Information](#i-information-message)
4. [Multi Information](#m-multi-information-message)
5. [Parameter](#p-parameter-message)
6. [Default Parameter](#q-default-parameter-message)

#### 'B': Flag Bits Message

:::note
This message must be the **first message** right after the header section, so that it has a fixed constant offset from the start of the file!
:::

```c
struct ulog_message_flag_bits_s {
  struct message_header_s header; // msg_type = 'B'
  uint8_t compat_flags[8];
  uint8_t incompat_flags[8];
  uint64_t appended_offsets[3]; // file offset(s) for appended data if appending bit is set
};
```

- `compat_flags`: compatible flag bits

>   - These flags indicate any information that is compatible with the existing ULog is in the log.

Author: It was explained a bit by Hamish here: https://github.com/PX4/PX4-user_guide/pull/1887#discussion_r882256016, but I am still confused as well :shrug:. Any contextual example would be much more helpful!

- `format` is a plain-text string with the following format: `message_name:field0;field1;`
  - There can be an arbitrary amount of fields (minimum 1), separated by `;`.

A `field` has the format: `type field_name`, or for an array: `type[array_length] field_name` is used (only fixed size arrays are supported).

A `type` is one of the [basic binary types](#data-types) or a `message_name` of another format definition (nested usage).
- A type can be used before it's defined.
  - e.g. The message `MessageA:MessageB[2] msg_b` can come before the `MessageB:uint_8[3] data`
- There can be arbitrary nesting but **no circular dependencies**
  - e.g. `MessageA:MessageB[2] msg_b` & `MessageB:MessageA[4] msg_a`

Some field names are special:
- `timestamp`: every [Subscription Message](#a-subscription-message) must include a timestamp field
  - Its type can be: `uint64_t` (currently the only one used), `uint32_t`, `uint16_t` or `uint8_t`.
  - The unit is always microseconds, except for in `uint8_t` where the unit is in milliseconds.
  - A log writer must make sure to log messages often enough to be able to detect **wrap-arounds** (when the timestamp overflows the data type and goes back to 0)

>   - Also, the log reader must handle wrap-arounds (and take into account dropouts).

Author: I added the 'optional' part, but the dropouts part is quite unclear to me still :shrug: (not sure how it works)

>   - Also, the log reader must handle wrap-arounds (and take into account dropouts).

Author: I edited it to make it more contextual :thinking:

  - The timestamp must always be monotonic increasing for a message series with the same `msg_id`.
- `_padding{}`: field names that start with `_padding` (e.g. `_padding[3]`) should not be displayed and their data must be ignored by a reader.
  - These fields can be inserted by a writer to ensure correct alignment.
  - If the padding field is the last field, then this field will not be logged, to avoid writing unnecessary data.
  - This means the `message_data_s.data` will be shorter by the size of the padding.
  - However the padding is still needed when the message is used in a nested definition.

#### 'I': Information Message

The Information message defines a dictionary type definition `key` : `value` pair for any information, including but not limited to Hardware version, Software version, Build toolchain for the software, etc.

```c
struct ulog_message_info_header_s {
  struct message_header_s header; // msg_type = 'I'
  uint8_t key_len;
  char key[key_len];

>   char value[header.msg_size-2-key_len]

Author: It was an mistake!

2. [Unsubscription](#r-unsubscription-message)
3. [Logged Data](#d-logged-data-message)
4. [Logged String](#l-logged-string-message)
5. [Tagged Logged String](#c-tagged-logged-string-message)
6. [Synchronization](#s-synchronization-message)
7. [Dropout Mark](#o-dropout-message)
8. [Information](#i-information-message)
9. [Multi Information](#m-multi-information-message)
10. [Parameter](#p-parameter-message)
11. [Default Parameter](#q-default-parameter-message)

#### `A`: Subscription Message

Subscribe a message by name and give it an id that is used in [Logged data Message](#d-logged-data-message).
This must come before the first corresponding [Logged data Message](#d-logged-data-message).

```c
struct message_add_logged_s {
  struct message_header_s header; // msg_type = 'A'
  uint8_t multi_id;
  uint16_t msg_id;
  char message_name[header.msg_size-3];
};
```

- `multi_id`: the same message format can have multiple instances, for example if the system has two sensors of the same type. The default and first instance must be 0.
- `msg_id`: unique id to match [Logged data Message](#d-logged-data-message) data. The first use must set this to 0, then increase it.

>   - The same `msg_id` must not be used twice for different subscriptions, not even after unsubscribing.

Reviewer 2: I don't see why it should not be possible to re-use after unsubscribing.

Author: Fixed!

In practice, we include the `msg_size` and `msg_type` in each message definitions as the first 2 members of different message definitions.

:::note
Messages below are prefixed with the single alphabet that corresponds to it's `msg_type`.
:::

### Definitions Section

The definitions section contains basic information such as software version, uORB Topic format, initial parameter values, and so on.

The message types in this section are:
1. [Flag Bits](#b-flag-bits-message)
2. [Format Definition](#f-format-message)
3. [Information](#i-information-message)
4. [Multi Information](#m-multi-information-message)
5. [Parameter](#p-parameter-message)
6. [Default Parameter](#q-default-parameter-message)

#### 'B': Flag Bits Message (`msg_type = 'B'`)

:::note
This message must be the **first message** right after the header section, so that it has a fixed constant offset from the start of the file!
:::

```c
struct ulog_message_flag_bits_s {
  uint16_t msg_size;
  uint8_t msg_type = 'B';
  uint8_t compat_flags[8];
  uint8_t incompat_flags[8];
  uint64_t appended_offsets[3]; ///< file offset(s) for appended data if appending bit is set
};
```

> - `compat_flags`: compatible flag bits

Reviewer: > How do we explain the compat_flags and incompat_flags more clearly? -> https://github.com/PX4/PX4-user_guide/pull/1887#issuecomment-1137202556

I don't find this confusing, perhaps because the concept is also used in mavlink.

Essentially there are two types of changes you might make to protocol or log or whatever - compatible or incompatible. 

An incompatible change is one that makes the content usuable - you simply have to throw away everything in the scope of the change.
- In mavlink this might be something "like" compression of a message - if a parser were to see a compression incompatibility bit set and it didn't know what that meant it would have to drop the message
- in the log this means you have to discard the whole log if you don't implement the feature indicated by an incompatible bit. This might mean the indicated extension is essential to using the log. I "guess" it could also be like the mavlink case, which means it makes the log unparsable. 

A compatible change is one that a parser can just ignore - it should continue to parse as normal - the output might not include some information that is in the log, but that isn't essential, and the parser can just go it's merry way.
For mavlink this is something like a bit that indicates a particular message is high priority. If a parser doesn't implement high priority parsing and the bit is appended to the end so that parsers that don't understand it won't even see it - then it is compatible. For the log this is just "default parameters message" - you can use a log even if you just ignore this message.

Anyway, leave to you to decide if  more needs to be said.

What this says is that if any of the `compat_flags` flags are set a parser that d

Author: Thank you for the explanation! It was applied in the latest commit

};
```

- `msg_size` is the size of the message in bytes without the header.
- `msg_type` defines the content, and is a single alphabet character.

:::note

> Message types described below are noted with the single alphabet `msg_type` characteristic in the beginning.

Author: Is this clear enough for the readers to recognize that the: `"'B': Flag Bits Message"` indicates that the `msg_type` of that message corresponds to the character `'B'`?

Because this is one of the first things I was so confused about the document, as it wasn't so clear what all those 'B', 'f', 'I', 'M', etc. meant :thinking:

Reviewer: No. We need to make that clear right up the top of the doc somewhere. I had same question and forgot to raise it.

Author: I added a [comment](https://github.com/PX4/PX4-user_guide/pull/1887#discussion_r876820169) about this, where the proposed change will expose the `msg_type` character clearly in the struct definition!

```

- `compat_flags`: compatible flag bits

>   - These flags indicate whether any information that is compatible with the ULog file format is in the log.

Reviewer: ```suggestion
  - These flags indicate the presence of features in the log file that are compatible with any ULog parser.
```

  - `compat_flags[0]`: *DEFAULT_PARAMETERS* (Bit 0): if set, the log contains [default parameters message](#q-default-parameter-message)

  The rest of the bits are currently not defined and must be set to 0.

>   These bits can be used for future ULog changes that are compatible with existing parsers. For example, adding a new message types, which can be indicated by setting one of the Bits, and the existing parsers can just ignore the new message type in the log.

Reviewer: ```suggestion
  These bits can be used for future ULog changes that are compatible with existing parsers.
  For example, adding a new message type can be indicated by defining a new bit in the standard, and existing parsers will ignore the new message type.
```

  It means parsers can just ignore the bits if one of the unknown bits is set.
- `incompat_flags`: incompatible flag bits.
  - `incompat_flags[0]`: *DATA_APPENDED* (Bit 0): if set, the log contains appended data and at least one of the `appended_offsets` is non-zero.
  
  The rest of the bits are currently not defined and must be set to 0.
  This can be used to introduce breaking changes that existing parsers cannot handle. For example, when an old ULog Parser that didn't have the concept of *DATA_APPENDED* reads the newer ULog, it would stop parsing the log as the log will contain out-of-spec messages / concepts.
  If a parser finds any of these bits set that isn't specified, it must refuse to parse the log.
- `appended_offsets`: File offset for appended data.
  If no data is appended, all offsets must be zero.

>   This can be used to reliably append data for logs that may stop in the middle of a message. For example, hardfault dumps in PX4.

Reviewer: ```suggestion
  This can be used to reliably append data for logs that may stop in the middle of a message.
  For example, crash dumps.
```

  A process appending data should do:
  - set the relevant `incompat_flags` bit

>   - set the first `appended_offsets` as the length of the log file without the appended data, as that is where the data would start

Reviewer 2: `set the first `appended_offsets` that is currently 0 to the length of the log file without the appended data, as that is where the new data will start`

Reviewer: ```suggestion
  - set the first `appended_offsets` that is currently 0 to the length of the log file without the appended data, as that is where the new data will start
```

>   - append any type of messages that are valid for the Data section.

Author: I am still unsure why it's only the 'first' element of the array `appended_offsets` that is used. Why do we have 3 of them as an array then?

Also, I would like to check if the phrase "length of the log file without the appended data, as that is where the data would start" is correct.

In PX4, it seems that the `incompat_flags` isn't set anywhere, except for having a check in reply module for replaying the log. So where is this `incompat_flags` actually set? In PX4? Or external source? @bkueng

It is possible that there are more fields appended at the end of this message in future ULog specifications.
This means a parser must not assume a fixed length of this message.
If the `msg_size` is bigger than expected (currently 40), any additional bytes must be ignored/discarded.

#### 'F': Format Message

> Format message defines a singe message name and it's inner fields in a single string.

Reviewer 2: `single`

Reviewer: ```suggestion
Format message defines a single message name and it's inner fields in a single string.
```

  - The unit is always microseconds, except for in `uint8_t` where the unit is in milliseconds.
  - A log writer must make sure to log messages often enough to be able to detect wrap-arounds and a log reader must handle wrap-arounds (and take into account dropouts).
  - The timestamp must always be monotonic increasing for a message series with the same `msg_id`.
- `_padding{}`: field names that start with `_padding` (e.g. `_padding[3]`) should not be displayed and their data must be ignored by a reader.
  - These fields can be inserted by a writer to ensure correct alignment.
  - If the padding field is the last field, then this field will not be logged, to avoid writing unnecessary data.
  - This means the `message_data_s.data` will be shorter by the size of the padding.
  - However the padding is still needed when the message is used in a nested definition.

#### 'I': Information Message

The Information message defines a dictionary type definition `key` : `value` pair for any information, including but not limited to Hardware version, Software version, Build toolchain for the software, etc.

```c
struct ulog_message_info_header_s {
  uint16_t msg_size;

>   uint8_t msg_type = 'I';

Author: So actually, it turns out that currently in PX4 the struct definitions don't use the `message_header_s` struct in the beginning of each message struct's definition. So the documentation was wrong (out of date).

So I propose putting in the actual implementation like so (only edited 2 messages to be in this format), what do you think? @hamishwillee

Reviewer 2: > So the documentation was wrong (out of date).

See my comment above - this isn't meant to document what PX4 exactly does.

Reviewer: @bkueng As an ordinary reader I did look at the introduction and assume that this document reflects what PX4 exactly does. Mostly because I assume this is a custom **PX4** file format definition. Not something abstract or from some external source.

We need to capture this in the introduction - ideally with some positioning for what this doc is/is not useful for in a PX4 context.

If it doesn't reflect PX4, where did this come from? And what use is the doc - just to get a general feel for the fields you might see?

Reviewer 2: Yes the file format is not tied to PX4. It could be used by any project with similar requirements.

PX4 follows the file format precisely, but it may use different struct namings, definitions, etc. Things that are not relevant for the on-disk file format.

  uint8_t key_len;
  char key[255];
};
```

- `key_len`: Length of the key value
- `key`: String with the `key` and `value` information combined: e.g. `char[5] sys_toolchain_ver9.4.0`
  - The `key` part has the format identical as the field in the [Format message](#f-format-message) (can also be a custom type), but consists of only a single field without ending `;`, eg. `char[5] sys_toolchain_ver`

>   - The `value` contains the data corresponding to the `key`, and is written right after the `key`, e.g. `9.4.0`.

Author: It was very confusing to explain how the key string of the struct differs from the concept of key : value of the information we are encompassing inside the key string.

Maybe worth changing the naming in the PX4's side to clarify what this field actually does? @bkueng

Author: PR here : https://github.com/PX4/PX4-Autopilot/pull/19675

`char[value_len] ver_hw` | Hardware version (board) |  "PX4FMU_V4"
`char[value_len] ver_hw_subtype` | Board subversion (variation)|  "V2"
`char[value_len] ver_sw` | Software version (git tag)|  "7f65e01"
`char[value_len] ver_sw_branch` | git branch |  "master"
`uint32_t ver_sw_release` | Software version (see below)|  0x010401ff
`char[value_len] sys_os_name` | Operating System Name |  "Linux"
`char[value_len] sys_os_ve`r | OS version (git tag) |  "9f82919"
`uint32_t ver_os_release` | OS version (see below) |  0x010401ff
`char[value_len] sys_toolchain` | Toolchain Name |  "GNU GCC"
`char[value_len] sys_toolchain_ver` | Toolchain Version |  "6.2.1"
`char[value_len] sys_mcu` | Chip name and revision |  "STM32F42x, rev A"
`char[value_len] sys_uuid` | Unique identifier for vehicle (eg. MCU ID) |  "392a93e32fa3"...
`char[value_len] log_type` | Type of the log (full log if not specified) | "mission"
`char[value_len] replay` | File name of replayed log if in replay mode | "log001.ulg"
`int32_t time_ref_utc` | UTC Time offset in seconds | -3600 |

:::note
`value_len` represents the data size of the `value`. This is described in the `key`.
:::

> - For PX4, the format of `ver_sw_release` and `ver_os_release` is: 0xAABBCCTT, where AA is **major**, BB is **minor**, CC is patch and TT is the **type**.

Reviewer 2: This does not only apply to PX4, please keep as it was

Reviewer: Revert
```suggestion
- The format of `ver_sw_release` and `ver_os_release` is: 0xAABBCCTT, where AA is **major**, BB is **minor**, CC is patch and TT is the **type**.
```

  uint8_t is_continued; // can be used for arrays
  uint8_t key_len;
  char key[key_len];

>   char value[header.msg_size-key_len]

Reviewer 2: This was correct with -2

Reviewer: Revert
```suggestion
  char value[header.msg_size-2-key_len]
```

| INFO       |      '6'     | Informational                        |
| DEBUG      |      '7'     | Debug-level messages                 |

#### 'S': Synchronization message

Synchronization message so that a reader can recover from a corrupt message by searching for the next sync message.

```c
struct message_sync_s {
  struct message_header_s header; // msg_type = 'S'
  uint8_t sync_magic[8];
};
```

* `sync_magic`: [0x2F, 0x73, 0x13, 0x20, 0x25, 0x0C, 0xBB, 0x12]

#### 'O': Dropout message

> Mark a dropout (lost logging messages) of a given duration in ms

Reviewer 2: ```suggestion
Mark a dropout (lost logging messages) of a given duration in ms.
```
